src/Microsoft.OData.Core/StreamExtensions.cs

```diff
@@ -29,5 +35,29 @@ public static byte[] ReadAllBytes(this Stream instream)
                 return memoryStream.ToArray();
             }
         }
+
+        /// <summary>
+        /// Asyncrhonously reads the data in the stream into a byte array.
+        /// </summary>
+        /// <param name="inStream">The stream to read data from.</param>
+        /// <returns>A task representing the asynchronous read operation.
+        /// On success, the result of the task will be the byte array containing
+        /// the read data.</returns>
+        public static async Task<byte[]> ReadAllBytesAsync(this Stream inStream)
+        {
+            Debug.Assert(inStream != null, "inStream != null");
+
+            if (inStream is MemoryStream)
+            {
+                byte[] contents = ((MemoryStream)inStream).ToArray();

```

Comment:
Given that in line 59 below you have `return memoryStream.ToArray()`, maybe here too you could have `return ((MemoryStream)inStream).ToArray()`